content/posts/gsoc-2024-auto-sync-sail/index.md

```diff
@@ -0,0 +1,151 @@
+
+In the context of Rizin, Lifting refers to transforming a machine code program written for any of the hardware architectures that Rizin understands (x86, RISC-V, 6502, etc...) to a Rizin-specific intermediate language called RzIL. 
+
+By doing this, Rizin’s developers can write generic analysis algorithms that manipulate RzIL instructions, and a generic VM that executes Rizin instructions, only once. Then, for each architecture that Rizin supports, a lifter that transforms machine code written for that architecture into RzIL is written, and we get all the analysis algorithms and VM execution capabilities “for free”. 
+
+In a nutshell, RzIL is the universal “Lingua Franca” for Rizin, like English is for Software Engineering.
+
+
+![IMAGE__WORLD_WITHOUT_RZIL](https://hackmd.io/_uploads/HyfLVzMHkg.png)
+Figure 1: Without RzIL, there is no smarter way to perform N operations for M assembly languages other than doing an NxM amount of work, implementing the N operations over and over again per each language/architecture.
+
+![IMAGE__WORLD_WITH_RZIL](https://hackmd.io/_uploads/HklxSGfHyg.png)
+Figure 2: With RzIL, the amount of work to support N operations for M architectures is N+M, the N operations are written exactly once for the intermediate language, then M transformers are written to lift each of the M architectures to the intermediate language.
+
+# For want of a disassembler
+
+So the original plan was to write the grey arrow in the figure above: a lifter from RISC-V machine code into RzIL. However, the first step in doing that is to “parse” RISC-V instructions from their binary form into a convenient data structure. We call that “parsing” step disassembly, or, more accurately, decoding.
+
+>> Side Note:  lots of people, when “disassembly” and “assembly” are mentioned, will probably think of the following diagram:
+![IMAGE__ASM_DISASM](https://hackmd.io/_uploads/Sy9mIfzHkl.png)
+This is not wrong for most purposes. However, in the context of this writeup it’s better to have the following and more detailed picture in mind:
+![IMAGE__ASM_DISASM_ENC_DEC](https://hackmd.io/_uploads/BySYvMfByx.png)
+In this writeup I’m more interested in the left-to-right flow: decoding from a binary to a structured (e.g. C struct) representation of the instruction, then assembling the structured representation of the instruction into a string form. Confusingly, sometimes “Disassembly” is used to include both Disassembly **and** Decoding, for example in Capstone the structured representation includes as a member its own toString serialization. It will often be clear from context what step is meant, and decoding is often far more important than disassembly. 
```

Comment:
```suggestion
In this writeup I’m more interested in the left-to-right flow: decoding from a binary to a structured (e.g. C struct) representation of the instruction, then assembling the structured representation of the instruction into a string form. Confusingly, sometimes “Disassembly” is used to include both Disassembly **and** Decoding, for example in Capstone the structured representation includes as a member its own `toString` serialization. It will often be clear from context what step is meant, and decoding is often far more important than disassembly. 
```